File: Dynamic-Programming/115-H-Distinct-Subsequences/distinct_subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m = len(t)
        n = len(s)
        OPT = [[0] * (n+1) for _ in range(m+1)]
        OPT[0][0] = 1

        for k in range(1, n+1):
            OPT[0][k] = 1

        for i in range(1, m+1):
            for j in range(1, n+1):
                if i>j:
                    continue
                if t[i-1] == s[j-1]:
                    OPT[i][j] = OPT[i-1][j-1] + OPT[i][j-1]
                else:
                    OPT[i][j] = OPT[i][j-1]
        
        return OPT[m][n]

File: Dynamic-Programming/115-H-Distinct-Subsequences/distinct_subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int n = s.length();
        int m = t.length();

        vector<vector<unsigned int>> OPT(m+1, vector<unsigned int>(n+1, 0));
        for (int j=0; j<n+1; j++) {
            OPT[0][j] = 1;
        }

        for (int i=1; i<m+1; i++) {
            for (int j=1; j<n+1; j++) {
                if (i>j) continue;

                if (s[j-1] == t[i-1])
                    OPT[i][j] = OPT[i-1][j-1] + OPT[i][j-1];
                else
                    OPT[i][j] = OPT[i][j-1];
            }
        }
        return OPT[m][n];
    }
};